src/addr2line.c

```diff
@@ -218,9 +218,72 @@ int addr2line__symbolize(const struct addr2line *a2l, long addr, struct a2l_resp
 	return cnt;
 }
 
-int addr2line__query_symbols(const struct addr2line *a2l, const char *compile_unit, struct a2l_resp *resp)
+int addr2line__query_symbols(const struct addr2line *a2l, const char *compile_unit, struct a2l_cu_resp *resp, int resp_size)
 {
 	int cnt = 0;
+	int err;
+
+	err = fprintf(a2l->write_pipe, "query_syms %s\n", compile_unit);
+	if (err <= 0) {
+		err = -errno;
+		fprintf(stderr, "Failed to get function names from compile unit(s): %d\n", err);
+		return err;
+	}
+	fflush(a2l->write_pipe);
+
+	while (true) {
+		char line[256];
+		if (fgets(line, sizeof(line), a2l->read_pipe) == NULL) {
+			err = -errno;
+			fprintf(stderr, "Failed to get functions from compile unit(s): %d\n", err);
+			return err;
+		}
+		if (line[0] == ':') {
+			if (line[1] == 'q') {
+				// :q is the last end of the result
+				break;
+			}
+			// Skip :e lines since we don't need filename so far
+			continue;
+		}
+		if (line[0] != ' ') {
+			fprintf(stderr, "Invalid format: %s\n", line);
+			return -1;
+		}
+
+		// |line| should be in the format of <spc><fuction><spc><address>
+
+		if (cnt > resp_size) {
+			// skip functions exceeding the size of the resp buffer
+			continue;
+		}
+
+		// Get function name
+		char *sep = strchr(line + 1, ' ');
+		if (sep == NULL) {
+			fprintf(stderr, "Invalid format: %s\n", line);
+			return -1;
+		}
+		*sep = 0;
+
+		if (sep - (line + 1) >= sizeof(resp->fname)) {
+			fprintf(stderr, "Function name is too long: %s\n", line + 1);
+			return -1;
+		}
+		strcpy(resp->fname, line + 1);
+
+		// Get address
+		char *addr_str = sep + 1;
+		if (addr_str[0] != '0' || addr_str[1] != 'x') {
+			fprintf(stderr, "Invalid address for function: %s %s\n", resp->fname, addr_str);
+			return -1;
+		}
+		addr_str += 2;
+		resp->address = (void*)strtoull(addr_str, NULL, 16);
```

Comment:
nit: space between `void` and `*`, but also please cast it through uintptr_t, because this can lead to compiler warnings on some architectures:

```
resp->address = (void *)(uintptr_t)strtoul(...);
```

you also don't need ull variant, sizeof(long) matches sizeof(void*) on any sane architecture